src/store.rs

```diff
@@ -25,11 +29,37 @@ pub fn get_missing_files(root: &str, wanted: &[String]) -> io::Result<Vec<String
             Ok(true) => None,
             Err(e) => Some(Err(e)),
         })
-    .collect()
+        .collect()
+}
+
+pub async fn put_file(root: &str, mut file: TempFile<'_>, hash: &str) -> io::Result<()> {
+    let temp_dir = tempdir_in(root)?;
+    let temp_path = temp_dir.path().join(hash);
+    file.persist_to(&temp_path).await?;
+    let alg = config::read_config(root)?.core.hash_algorithm;
+    let content = fs::read_to_string(&temp_path)?;
+    let valid_hash = hash::hash_data(content, alg);
+    if hash != valid_hash {
```

Comment:
Something just occurred to me; isn't it the case that the local user could legitimately be using a different hash algorithm to the remote? In which case we should include the algorithm in the request as well?

Reply:
yes, this is theoretically possible but there's heaps of support required to work out what would happen so won't happen practically for a while. The hash should be validated against the hash string (so parse the algorithm out of the hash and validate that, store the file in a path that matches that key, then *also* hard link a copy of our preferred algorithm's hash after reshashing). See https://github.com/mrc-ide/outpack/pull/77/files for some comments on that.

For now, consider checking if core.hash_algorithm matches the provided one and erroring if not compatible? This will work for all practical cases and be nicely future proof